keps/sig-node/3322-add-maxRestartTimes-to-podSpec/README.md

```diff
@@ -0,0 +1,981 @@
+bogged down.
+-->
+
+#### Story 1
+
+When running with a pod, some system errors or program internal errors
+encountered, which causes the application restarting infinitely. This is
+unrecoverable and wasteful.
+
+### Story 2
+
+For batch Job specifically, now a Job provides a single backoffLimit for all derived pods,
+which may cause the entire Job failed and clean up all the resources because of one of the pods
+restarts for more than the `spec.backoffLimit`. To solve the problem, the Batch WG incubates
+a new KEP [Backoff Limit Per Index For Indexed Jobs](https://github.com/kubernetes/enhancements/issues/3850)
+which introduces a new field `backoffLimitPerIndex` to control the max restart times per Pod.
+
+It's actually what we want to implement here, by sinking the capacity of `backoffLimitPerIndex` to pod level,
+it helps to reduce the maintenance costs of the Job API.
+
+Note: currently, `backoffLimitPerIndex` only supports `restartPolicy=Never` initially, but as the KEP states, it will depend
+on this feature to support onFailure restartPolicy in the future, see [chapter](https://github.com/kubernetes/enhancements/blob/ab54c14de86af3bcc76b45a4c9194bc47f6c9a62/keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs/README.md#support-backofflimitperindex-when-restartpolicyonfailure).
+This reminds me whether we should widen the _max_restart_times_ to other restart policies like `Never`, see discussion below.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+#### Discussions
+
+The Batch WG incubates a new KEP [Backoff Limit Per Index For Indexed Jobs](https://github.com/kubernetes/enhancements/issues/3850),
+which introduces a new field `backoffLimitPerIndex`, then each pod of the Job will have an independent backOffLimit,
+but initially it only supports `restartPolicy=Never`, maybe we can widen the _max_restart_times_ to support other restart policies,
+like `Never`, then the Job API no longer needs to consider these logics anymore. Currently, it uses an annotation `batch.kubernetes.io/job-index-failure-count`
+to record the retires number, which looks not elegant.
+
+Pros:
+  * BackoffLimitPerIndex can reuse this functionality and no longer need to consider the restart times per index.
+  Specifically, it can avoid to use the annotation anymore, and works at a high level control by watching the pod status.
+
+Cons:
+  * If we support max restart times with `restartPolicy=Never`, it semantically means _we can never
+  restart the pod until we met the restart times. Then there's little difference with `restartPolicy=OnFailure`.
+
+#### Caveats
+
+- When a Job's template is set with restartPolicy=OnFailure, and `backoffLimit` is greater than `maxRestartTimesOnFailure`,
```

Comment:
I guess you are talking about the case when parallelism=1.

I'm not liking these semantics, even if they are new. The backoffLimit in Job is already not accurate when restartPolicy=OnFailure https://kubernetes.io/docs/concepts/workloads/controllers/job/#pod-backoff-failure-policy and https://github.com/kubernetes/kubernetes/issues/109870

Note that when Pod is marked as Failed (as it will be when the Pod reaches `maxRestartTimes`), the logic would count the number of failures as 1, regardless of the number of container restarts. I know, it's ugly, but it's the legacy behavior. This might not give the desired outcome for users of Job.

What if when if the job controller observes the failure reason/condition for "reached max restarts", it would add this to the number of failures and count towards backoffLimit. You need to consider finalizers, but maybe it's doable?

We could also consider not allowing users of Job to set `maxRestartTimes` and let the job controller add it, in behalf of the user, when creating a Pod. But how does a user tell the Job API whether to use the old counting method or use `maxRestartTimes`?

So, overall, I think this section needs to consider some alternatives to deal with Job backoffLimit, for different levels of parallelism.

Reply:
cc @mimowo @soltysh

Reply:
I have a different proposal - when a user specifies `maxRestartTimes`, then `backoffLimit` should only count pod when they are `Failed`, instead of checking individual containers. 

This puts a clear separation of concerns between kubelet and Job controller, eliminating the issue https://github.com/kubernetes/kubernetes/issues/109870.

Reply:
> I have a different proposal - when a user specifies maxRestartTimes, then backoffLimit should only count pod when they are Failed, instead of checking individual containers.

I think this is the easiest solution. But would that be useful to users of Job? That I'm not sure about.

> So, another proposal would be that Job controller sets maxRestartTimes but based on the backoffLimitPerIndex (and the annotation counting how many failures happened before recreation).

In the case of backoffLimitPerIndex it's much easier. The job controller would create the first Pod with full maxPodRestarts. If the Pod is preempted/evicted/etc, the replacement pod has a lower maxPodRestarts.

I'm not too sure what do do with regular backoffLimit... Maybe going with the easiest solution above is the best we can do?

Reply:
Actually, with the approach both `backoffLimit` and `backoffLimitPerIndex` can be implemented very similarly to as currently when `restartPolicy=Never`. We only count failed pods (not container restarts), we don't even need the annotation. Simply, user sets `maxRestartTimes` once the pod reaches the `Failed` phase it is counted towards the backoff limits.

Reply:
Whether `backoffLimit` and `backoffLimitPerIndex` are mutually exclusive? I guess they are. Then `backoffLimit` and `maxRestartTimes` should be mutually exclusive similarly. If `backoffLimitPerIndex` counts the restart times on individual containers(indirectly via maxRestartTimes), then `backoffLimitPerIndex` should also.

Reply:
OK, this is not real. see https://github.com/kubernetes/enhancements/pull/3967/files#diff-4908561f5b948cdeee23dd68a5a9d77208c89394bb2f05b898a8d679802a8e46R1313

Reply:
Revisit the reasons we defer to make `backoffLimit` and `backoffLimitPerIndex` mutually exclusive

> There is no way to control downgrade, as the value of `backoffLimit` would always default to 6. 
> Also, old API clients may error trying to read or modify Job objects with backoffLimit=nil.

Can you explain more about this? @mimowo Thanks a lot. I think make them mutually exclusive is meaningful.

Reply:
The intention is that when `backoffLimitPerIndex` is used, the `backoffLimit` is not functional. However, we set `backoffLimit` to max int 32, rather than not set it at all. The reasons are:
- if it remained `nil`, then some older API clients may fail to perform updates
- setting `backoffLimit` in may allow for controlled downgrade, by setting it to `backoffLimitPerIndex*completions*expectedFailureRate`

Reply:
> Whether backoffLimit and backoffLimitPerIndex are mutually exclusive? I guess they are. Then backoffLimit and maxRestartTimes should be mutually exclusive similarly.

I *think* if `maxRestartTimes` is specified, then I would suggest that `backoffLimit` and `backoffLimitPerIndex` are functional, but count only Failed pods.

Reply:
Taking a step back, maybe what @mimowo proposes is acceptable for `backoffLimit`, but not for `backoffLimitPerIndex`

Reply:
Indeed the number of restarts when `backoffLimitPerIndex` is specified along with `maxRestartTimes` would be: `maxRestartTimes * (backoffLimitPerIndex+1)`. I think there are use cases for `backoffLimitPerIndex > 0` to account for pod eviction, say due to exceeding ActiveDeadlineSeconds timeout, or other eviction / preemption scenarios. It also makes the semantic well defined, and analogous to `backoffLimit`.

Reply:
I like the idea that the only difference between `backoffLimit` and `backoffLimitPerIndex` is that one counts failures globally, and the other per index.

Reply:
I'm supportive of the idea of a global `backoffLimit` and local `maxRestarts`, which allows granular control over your job. For Indexed Jobs where you set `backoffLimitPerIndex` it seems reasonable that we'll copy that value into that new field `maxRestarts` to get rid of the extra counting we currently have from job controller. On the other hand we clearly excluded `OnFailure` from that other KEP, but the discussions here are around having a global mechanism, so how this affects `backoffLimitPerIndex` are still to be decided once we have a crisp idea how this proposals finishes. https://github.com/kubernetes/enhancements/issues/3850 is currently in alpha, I'm inclined to block GA of it until we have this feature alpha or ideally beta.

Reply:
> ndeed the number of restarts when `backoffLimitPerIndex` is specified along with `maxRestartTimes` would be: `maxRestartTimes * (backoffLimitPerIndex+1)`

The problem is that this is not always accurate. If a Pod is evicted before it completes its `maxRestartTimes`, then the equation no longer applies. So overall I think this makes the feature highly unpredictable.

I like the idea of making #3850 GA depend on beta for #3322. But I really see it as the job controller using the field `maxRestarts`, rather than letting the user set it up.

Reply:
In practice, what this means for the purpose of #3322 going to alpha, is that we should have a validation rule in Job that forbids users from setting `maxRestarts` for Indexed Jobs (or something along those lines).

Reply:
I'm not convinced about blocking graduation of #3850, which is technically scoped to `Never` so there is no direct interaction, unless we want to extend its scope.

I see these approaches:
1. limit #3850 to `Never`, and support `maxRestarts` with `backoffLimitPerIndex` and `backoffLimit` when `OnFailure` in a dedicated KEP - that KEP could also enable it for podFailurePolicy
2. support `maxRestarts` with `backoffLimitPerIndex` when `OnFailure` (whatever the solution), putting aside `backoffLimit` under #3850 
3. support `maxRestarts` with `backoffLimitPerIndex` when `OnFailure` (whatever the solution), along with `backoffLimit`

My preference is for (1.) to keep KEPs small and graduating fast, but I'm ok with (2.). My preference for (1.) is that when approaching (2.) we may scope creep to (3.), and possibly pod failure policy. 

Thus, I feel like there is room for a dedicated KEP focused on enabling `OnFailure` with `maxRestarts` for `backoffLimitPerIndex`, `backoffLimit` and pod failure policy. The need for pod failure policy was also discussed here: https://github.com/kubernetes/enhancements/issues/3329#issuecomment-1569239302.